ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/QueryFilterNode.java

```diff
@@ -447,10 +444,49 @@ public Void visitComparisonExpression(
         final Object key = resolveKey(other, col.get(), metaStore, ksqlConfig, node);
         keyContents[col.get().index()] = key;
         seenKeys.set(col.get().index());
+        operatorType = node.getType();
```

Comment:
Interesting. Thanks for pointing this out, @AlanConfluent . You're saying that currently, if we have a multi-column key, we can't do table-scan queries today? It seems like we ought to have a test for that kind of thing. It also seems like that restriction isn't strictly necessary.

Reply:
FWIW, I think multi-column keys are already excluded from the optimization, since @patrickstuedi restricted the key domain to just plain String keys. It's worth checking that multi-column keys do indeed fall back to table scans.

Reply:
@vvcephei @AlanConfluent yes we currently fall back to table scan for multi-range expressions, there are a set of tests in the rest validation tests "pull-queries-with-scan.json"

https://github.com/confluentinc/ksql/pull/7993/files#diff-ace0366edb6fa80100210b650c4b01793f397c85d352a9683d59851a70637b3b

Reply:
> You're saying that currently, if we have a multi-column key, we can't do table-scan queries today? It seems like we ought to have a test for that kind of thing. It also seems like that restriction isn't strictly necessary.

I didn't meant to imply that.  Table scan should work for this today. The validator above will check to see if any operation isn't an "=" and potentially make it a scan if it is.  Currently, only if they're all "=" will it skip a table scan and do a key lookup.  To handle range queries, this all still holds true, but we can additionally carve out the case of a single key column with range operator from the cases that were being handled by table scan, and now handle that by range query.

Reply:
I believe the current code, if you have multiple key columns, will not behave correctly for a query like this:
- `KEY1 < 10 AND KEY2 = 5`

operatorType will be set to the latest operator seen in the AST traversal, which is equals.  It will, based on that, assume they're all equality, which was a simplifying assumption we could make before because we only allowed equality, but now isn't quite accurate.  I think the current code would just do an equality lookup rather than range operator on one column.  If we see this intermixing of operators, we should just fall back on table scan.

To keep track of that, I think we need to keep track of all of the operators.  Only if they're all "=" can the lookup operator be equal.  I think intermixing is a table scan and if it's a single column key, then range operators can result in lookup operators that are range.  My comments above were meant to say that even if we have all "<" for a multi column key, we wouldn't want to do a range scan operation since it wouldn't make sense.

> At the moment multi key columns only work for equality, so other operators should throw an error in that case.

I should have said fall back on table scan.